src/Datadog.Trace.ClrProfiler.Managed/AutoInstrumentation/RabbitMQ/BasicPublishIntegration.cs

```diff
@@ -0,0 +1,96 @@
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Tagging;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.RabbitMQ
+{
+    /// <summary>
+    /// RabbitMQ.Client BasicPublish calltarget instrumentation
+    /// </summary>
+    [InstrumentMethod(
+        Assembly = "RabbitMQ.Client",
+        Type = "RabbitMQ.Client.Framing.Impl.Model",
+        Method = "_Private_BasicPublish",
+        ReturnTypeName = ClrNames.Void,
+        ParametersTypesNames = new[] { ClrNames.String, ClrNames.String, ClrNames.Bool, RabbitMQConstants.IBasicPropertiesTypeName, ClrNames.Ignore },
+        MinimumVersion = "3.6.9",
+        MaximumVersion = "6.*.*",
+        IntegrationName = RabbitMQConstants.IntegrationName)]
+    public class BasicPublishIntegration
+    {
+        private const string Command = "basic.publish";
+
+        private static readonly string[] DeliveryModeStrings = { null, "1", "2" };
+
+        private static readonly Vendors.Serilog.ILogger Log = DatadogLogging.GetLogger(typeof(BasicPublishIntegration));
+
+        /// <summary>
+        /// OnMethodBegin callback
+        /// </summary>
+        /// <typeparam name="TTarget">Type of the target</typeparam>
+        /// <typeparam name="TBasicProperties">Type of the message properties</typeparam>
+        /// <typeparam name="TBody">Type of the message body</typeparam>
+        /// <param name="instance">Instance value, aka `this` of the instrumented method.</param>
+        /// <param name="exchange">Name of the exchange.</param>
+        /// <param name="routingKey">The routing key.</param>
+        /// <param name="mandatory">The mandatory routing flag.</param>
+        /// <param name="basicProperties">The message properties.</param>
+        /// <param name="body">The message body.</param>
+        /// <returns>Calltarget state value</returns>
+        public static CallTargetState OnMethodBegin<TTarget, TBasicProperties, TBody>(TTarget instance, string exchange, string routingKey, bool mandatory, TBasicProperties basicProperties, TBody body)
+            where TBasicProperties : IBasicProperties, IDuckType
+            where TBody : IBody // Versions < 6.0.0: TBody is byte[] // Versions >= 6.0.0: TBody is ReadOnlyMemory<byte>
```

Comment:
@tonyredondo in our CallSite instrumentation, due to the changing type of the body parameter, we have two BasicPublish integrations: One for versions < 6.0.0 and one integration for versions >= 6.0.0. Will it be performant to combine them this way in our CallTarget instrumentation? Or should I implement them separately again?

Reply:
It's ok keeping both combined. First I think both versions of this library cannot be loaded to the same app and context. But also the caller in this case it's a delegate with the signature:

```csharp
internal delegate CallTargetState InvokeDelegate(TTarget instance, TArg1 arg1, TArg2 arg2, TArg3 arg3, TArg4 arg4, TArg5 arg5);
```

stored in a class with the signature:

```csharp
internal static class BeginMethodHandler<TIntegration, TTarget, TArg1, TArg2, TArg3, TArg4, TArg5>
```

Being `TArg5 == TBody` if the `TBody` type changes then the delegate will be stored in another class type and it will be isolated.

Reply:
>  I think both versions of this library cannot be loaded to the same app and context

Ha, you'd be surprised by the things we've seen...